### PR TITLE
Fix handling of empty values on documentation generation

### PR DESCRIPTION
**Pull request by penev92**
I noticed something we missed in #19948 - we never had `null` as a default value for a trait or a weapon field. The reason being how `FieldSaver.SaveField()` works.
Since the distinction between an empty string and `null` can be important for consumers of the documentation, I added a quick "hack" to fix it. Also no values, including empty strings, make sense on fields with `FieldLoader.RequireAttribute`, so I changed those to `null`. No update rules needed either.

I suggest reviewing with whitespace changes hidden - https://github.com/OpenRA/OpenRA/pull/20250/files?w=1

### CODE REVIEW

File: OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs [hunk 1]
```diff
@@ -26,7 +26,7 @@ public class ClonesProducedUnitsInfo : ConditionalTraitInfo, Requires<Production
 
 		[FieldLoader.Require]
 		[Desc("e.g. Infantry, Vehicles, Aircraft, Buildings")]
-		public readonly string ProductionType = "";
+		public readonly string ProductionType = null;
```
Comment (PunkPun) [inline OpenRA.Mods.Cnc/Traits/Buildings/ClonesProducedUnits.cs:29]:
```suggestion
		public readonly string ProductionType;
```

Couldn't we just do this instead? It's null by default

File: packaging/format-docs.py [hunk 2]
```diff
@@ -83,7 +83,7 @@ def format_docs(version, collectionName, types):
 
                         print(f'| {prop["PropertyName"]} | {defaultValue} | {prop["UserFriendlyType"]} | {prop["Description"]} |')
                     else:
-                        print(f'| {prop["PropertyName"]} | {prop["DefaultValue"]} | {prop["UserFriendlyType"]} | {prop["Description"]} |')
+                        print(f'| {prop["PropertyName"]} | {prop["DefaultValue"] or ""} | {prop["UserFriendlyType"]} | {prop["Description"]} |')
```
Comment (PunkPun) [inline packaging/format-docs.py:86]: Why is this needed? When I revert this change the end result is identical

File: OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs [hunk 3]
```diff
@@ -59,31 +59,42 @@ static string GenerateJson(string version, IEnumerable<Type> traitTypes, ObjectC
...
-								.Where(a => a.AttributeType.Name != nameof(DescAttribute) && a.AttributeType.Name != nameof(FieldLoader.LoadUsingAttribute))
-								.Select(a =>
-								{
-									var name = a.AttributeType.Name;
-									name = name.EndsWith("Attribute") ? name.Substring(0, name.Length - 9) : name;
+							var typeInstance = objectCreator.CreateBasic(type);
+							var defaultValue = FieldSaver.SaveField(typeInstance, fi.Field.Name).Value.Value;
 
-									return new
+							// HACK: FieldSaver always produces an empty string if there is no value, but it can be important to distinguish between "" and null.
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs:67]: Would adding a `replaceNullWithEmptyString: false` argument to `FieldSaver.SaveField` solve the disagreements over this?

**Comment by PunkPun** [comment]
Is there a difference for the end user between an empty string an an uninitiated string? If not, then we don't need to document it. 

If there indeed isn't a difference we could instead choose one and apply a code style rule to all strings, forcing the use of either empty or uninitialised strings.

File: OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs (hunk 3 above)
Comment (penev92) [reply OpenRA.Mods.Common/UtilityCommands/ExtractTraitDocsCommand.cs:67]: Something like that could probably work, yes, or perhaps going directly for `FieldSaver.FormatValue(object v)` instead of `FieldSaver.SaveField()`.

File: packaging/format-docs.py (hunk 2 above)
Comment (penev92) [reply packaging/format-docs.py:86]: When you actually have `null` values in the generated JSON (which happens for sequences), the Python script serializes that as "None", hence this check to not show anything in the Markdown.

**Comment by penev92** [comment]
I don't see the need for the question here as I already tried explaining multiple times on Discord and was ignored *(I understand that this is a simple PR bump)*. People just decided they don't care about a difference there. So I'm just going to drop the changes here and move on to the next PR.
#whatever #minimalEffort #whoCares

**Comment by PunkPun** [comment]
> People just decided they don't care about a difference there. 

The problem was that you didn't mention a single difference, the question was never answered!

**Comment by PunkPun** [comment]
I think it's reasonable to ask for usecases / explanation before introducing a HACK and from it judge if a better solution could be used instead. Instaed of an answer you basically just said." it's better, do you not trust me?"

**Comment by PunkPun** [comment]
And then just started ignoring the subject until I posted it on github
